include/gc/Analysis/MatmulConfigAnalysis.h

```diff
@@ -10,64 +10,88 @@
+#include "mlir/Dialect/DLTI/DLTI.h"
 #include "mlir/Dialect/Linalg/IR/Linalg.h"
-#include <cstring>
+#include "mlir/Interfaces/DataLayoutInterfaces.h"
 
 namespace mlir {
 namespace gc {
 
 using namespace mlir;
 
-// A mock for the taget information
-// TODO: replace it with upstream hardware description model
 struct SystemDesc {
-
-  static int getPositiveIntFromStr(char *str, int defaultValue = 1) {
-    if (!str || strlen(str) == 0 || str[0] > '9' || str[0] < '0') {
-      return defaultValue;
-    }
-    auto val = std::stoi(str);
-    return val > 0 ? val : defaultValue;
-  }
-
   // get runtime OMP_NUM_THREADS
   uint32_t getNumThreads() {
-    char *numThreads = getenv("OMP_NUM_THREADS");
-    return getPositiveIntFromStr(numThreads, 1);
+    DataLayout layout = DataLayout(module);
+    MLIRContext *ctx = module->getContext();
+    std::optional<Attribute> numThreads = layout.getDevicePropertyValue(
+        Builder(ctx).getStringAttr("CPU" /* device ID*/),
+        Builder(ctx).getStringAttr("num_threads"));
+    if (numThreads && dyn_cast<IntegerAttr>(*numThreads)) {
+      return cast<IntegerAttr>(*numThreads).getInt();
+    }
+    return 1;
   }
   // get cache size by cacheLevel
   size_t getCacheSize(uint8_t cacheLevel) {
+    DataLayout layout = DataLayout(module);
+    MLIRContext *ctx = module->getContext();
+
     if (cacheLevel == 1) {
-      char *cacheSize = getenv("L1_CACHE_SIZE");
-      return getPositiveIntFromStr(cacheSize, 0);
+      std::optional<Attribute> cacheSize = layout.getDevicePropertyValue(
+          Builder(ctx).getStringAttr("CPU" /* device ID*/),
+          Builder(ctx).getStringAttr("L1_cache_size_in_bytes"));
+      if (cacheSize && dyn_cast<IntegerAttr>(*cacheSize)) {
+        return cast<IntegerAttr>(*cacheSize).getInt();
+      }
     } else if (cacheLevel == 2) {
-      char *cacheSize = getenv("L2_CACHE_SIZE");
-      return getPositiveIntFromStr(cacheSize, 0);
+      std::optional<Attribute> cacheSize = layout.getDevicePropertyValue(
+          Builder(ctx).getStringAttr("CPU" /* device ID*/),
+          Builder(ctx).getStringAttr("L2_cache_size_in_bytes"));
+      if (cacheSize && dyn_cast<IntegerAttr>(*cacheSize)) {
+        return cast<IntegerAttr>(*cacheSize).getInt();
+      }
     } else if (cacheLevel == 3) {
-      char *cacheSize = getenv("L3_CACHE_SIZE");
-      return getPositiveIntFromStr(cacheSize, 0);
+      std::optional<Attribute> cacheSize = layout.getDevicePropertyValue(
+          Builder(ctx).getStringAttr("CPU" /* device ID*/),
+          Builder(ctx).getStringAttr("L3_cache_size_in_bytes"));
+      if (cacheSize && dyn_cast<IntegerAttr>(*cacheSize)) {
+        return cast<IntegerAttr>(*cacheSize).getInt();
+      }
     }
     return 0;
   }
 
   // get the maximum vector length in bits
   size_t getMaxVectorLength() {
-    char *maxVectorLanes = getenv("MAX_VECTOR_LENGTH");
-    return getPositiveIntFromStr(maxVectorLanes, 512);
+    DataLayout layout = DataLayout(module);
```

Comment:
Seems we can also add `layout` as a private member and only create once at instantiation stage.

Reply:
Thanks for the advice! Has replaced `moduleOp` with `layout` and `context` as private member